DS-3053 Shrink widget for wordbag and list of items

jrwishart opened the pull request:
The create text analysis widget has its widget code improved by 
reducing the duplication of styling code in CSS. The class names 
are reduced by changing from integer labelling to alphanumeric 
labelling. Tokens are shown via the after selector in css to reduce 
duplication. Hover text also tidied with only variants not the same
as the selected token being shown.

- DS-3023 Move common table props to CSS
- DS-3023 Condense styling in css
- DS-3023 Remove tokens from span element
- DS-3023 Change class name indices to different base
- DS-3023 Use anchor instead of span
- DS-3023 Tidy up tooltip text
- DS-3023 Separate style and content classes
- DS-3023 Make unique placeholders more efficient
- DS-3023 Remove quotes from examples
- DS-3053 Prevent incomplete final line warning [revdep skip]
- DS-3053 Start style class names with s
- DS-3053 Add logical check to token placeholders
- DS-3053 Amend css to support ListCat and wordBag [revdep skip]
- DS-3053 Handle unclassified tokens properly [revdep skip]
- DS-3053 Version bump
- DS-3053 Stop wildcard matching the summary class
- fixup! DS-3053 Amend css to support ListCat and wordBag [revdep skip]

JustinCCYap commented on R/createtextanalysiswidget.R:
```diff
@@ -99,47 +100,83 @@ HighlightNGrams <- function(n.grams, text, subs, category.examples,
+    # Create colours
+    potential.colours <- setAlpha(col0, 0.5)
+    # Make mapping for the colour index
+    color.index <- 1:n %% n.col
```
This line and the one below can be replaced with `color.index <- 0:(n - 1) %% n.col + 1`

jrwishart replied:
Thanks, fixed in https://github.com/Displayr/flipFormat/pull/22/commits/d602ebbbae522eb2f9fa5b015b1d1e0953456dac

JustinCCYap commented on R/createtextanalysiswidget.R:
```diff
@@ -99,47 +100,83 @@ HighlightNGrams <- function(n.grams, text, subs, category.examples,
+    # Replace the 0 remainder with length of number of colours
+    color.index[color.index == 0] <- n.col
+    # Remove Unclassified tokens since they get separate styling
+    unclassified <- which(n.grams[[1]] == "UNCLASSIFIED")
```
You could remove the `which` and replace `length` with `any` and `-unclassified` with `!unclassified`?

jrwishart replied:
Could definitely do that. The rationale was there will only be one unclassified token and the above approach would give a single index element instead of a logical of size of n.grams. Happy to change if it improves readability?

JustinCCYap replied:
Ah I see, it only appears once. I'm ok with the way it is. Generally if it appears multiple times I would avoid using which

JustinCCYap commented on R/createtextanalysiswidget.R:
```diff
@@ -99,47 +100,83 @@ HighlightNGrams <- function(n.grams, text, subs, category.examples,
+    # Create colours
+    potential.colours <- setAlpha(col0, 0.5)
+    # Make mapping for the colour index
+    color.index <- 0:n %% n.col + 1
```
This should be `color.index <- 0:(n - 1) %% n.col + 1`

jrwishart replied:
oops, thanks for picking that up. I changed that too quickly. I'll fix it again

JustinCCYap commented:
Why was the quantity added in the tooltips? I'd prefer if it wasn't there. The examples are limited in quantity anyway, so I don't see the need to tell the user how many there are. And it looks weird to me for lines to start with numbers. TBH I prefer the original way of showing the tooltips, but I guess I am biased since I created it.

![image](https://user-images.githubusercontent.com/13842380/91241242-abf85000-e787-11ea-833e-26d528d1669f.png)

JustinCCYap commented on R/utilities.R:
```diff
@@ -178,5 +178,23 @@ hasTextFromIndexOnwards <- function(labels, index, non.word.chars)
+    out <- ""
+    while (x)
+    {
+        r <- as.character(x %% base)
```
Why is it necessary to name `possible.chars`? Why not just define `r <- x %% base + 1`

jrwishart replied:
Thanks for picking this up. Similar to the index vector used in the `HighlightNGrams` function. Fixed in https://github.com/Displayr/flipFormat/pull/22/commits/e5c0b0252a100b1f60cf3c078905a93f528a496a

jrwishart commented on R/utilities.R:
```diff
@@ -178,5 +178,23 @@ hasTextFromIndexOnwards <- function(labels, index, non.word.chars)
+# valid css identifier name seem to be able to use base 64 (0-9,a-z,A-Z,- and _)
+# see for example the first bullet point at https://www.w3.org/TR/CSS21/syndata.html#characters
+# However, HTML class names seem to be case insensitive (so "wA" would be the same as "wa")
+#' @param x Integer vector, typically a sequence from 1:n that is to be converted to base64
```
That is a typo. Fixed the documentation in https://github.com/Displayr/flipFormat/pull/22/commits/e5c0b0252a100b1f60cf3c078905a93f528a496a

jrwishart commented:
> Why was the quantity added in the tooltips? I'd prefer if it wasn't there. The examples are limited in quantity anyway, so I don't see the need to tell the user how many there are. And it looks weird to me for lines to start with numbers. TBH I prefer the original way of showing the tooltips, but I guess I am biased since I created it.
> 
> ![image](https://user-images.githubusercontent.com/13842380/91241242-abf85000-e787-11ea-833e-26d528d1669f.png)

I couldn't reproduce that seemingly bad tooltip with an example of just "h" but I have restored the old tooltip layout with the exception that Example(s) is now replaced with Example or Examples depending on 1 vs >1 example in each case.

JustinCCYap commented:
> > Why was the quantity added in the tooltips? I'd prefer if it wasn't there. The examples are limited in quantity anyway, so I don't see the need to tell the user how many there are. And it looks weird to me for lines to start with numbers. TBH I prefer the original way of showing the tooltips, but I guess I am biased since I created it.
> > ![image](https://user-images.githubusercontent.com/13842380/91241242-abf85000-e787-11ea-833e-26d528d1669f.png)
> 
> I couldn't reproduce that seemingly bad tooltip with an example of just "h" but I have restored the old tooltip layout with the exception that Example(s) is now replaced with Example or Examples depending on 1 vs >1 example in each case.

The "h" example tooltip is because letters are passed to `CreateTextAnalysisWidget` in the unit test

JustinCCYap commented:
Could you have a look at why coverage decreased and try increasing it?

https://coveralls.io/builds/33033318/source?filename=R%2Fcreatetextanalysiswidget.R

jrwishart commented:
Added unit tests for the functions I added in master in the commits https://github.com/Displayr/flipFormat/commit/736df5118b8ddedfa018197f3f8246db65d5c75b and 4cf6a0da60db15bf3afbbd8c09d95483246d0fd1 to alleviate the printing issue of Text widgets used in other cases. The coverage has now been restored.

JustinCCYap commented:
For some reason I get the error "R server died" on text analysis outputs when running this branch on reusdev:
https://app.displayr.com/Dashboard?project_id=-205098#page=f3af2f97-cda8-4d94-9923-411942deafac
When I uninstall the branch, everything becomes fine again.

jrwishart commented:
I should have mentioned. This branch is actually DS-3053-2. DS-3053 has the minifyHTML implemented. It calls V8 and causes issues on the dev server. DS-3053-2 doesn't use the minifier and works fine on dev
